src/kvtree_util.c

```diff
@@ -225,3 +230,130 @@ int kvtree_util_get_ptr(const kvtree* hash, const char* key, void** val)
 
   return rc;
 }
+
+/*
+ * Given a prefix passed to a kvtree_write_gather() (like "/tmp/rank2file",
+ * read all the rank data from all the scattered subfiles, and construct a
+ * kvtree in 'data'.  'data' will look something like:
+ *
+ *       26
+ *         FILE
+ *           ckpt.1/rank_26.ckpt
+ *       24
+ *         FILE
+ *           ckpt.1/rank_24.ckpt
+ *       25
+ *         FILE
+ *           ckpt.1/rank_25.ckpt
+ *       18
+ *         FILE
+ *           ckpt.1/rank_18.ckpt
+ *       ...
+ *
+ *  ... with each top-level element being the rank number.
+ */
+int kvtree_read_scatter_single(const char* prefix, kvtree* data)
+{
+  int rc = KVTREE_SUCCESS;
+  unsigned long level, expected_ranks, actual_ranks = 0;
+  char *prefix_dir, *prefix_dir_copy = NULL, *prefix_file, *prefix_file_copy = NULL;
+  char tmp[PATH_MAX], pattern[PATH_MAX];
+  kvtree *subfile_tree = NULL, *subfile_rank_tree = NULL;
+  kvtree *final_tree = kvtree_new();
+  regex_t regex;
+
+  /*
+   * Look at the high level kvtree and get the total number of ranks to
+   * expect.
+   */
+  rc = kvtree_read_file(prefix, final_tree);
+  if (rc != KVTREE_SUCCESS) {
+    kvtree_free(&final_tree);
+    return rc;
+  }
+
+  rc = kvtree_util_get_unsigned_long(final_tree, "RANKS", &expected_ranks);
+  if (rc != KVTREE_SUCCESS) {
+    kvtree_free(&final_tree);
+    return rc;
+  }
+  kvtree_delete(&final_tree);
+
+  /* Start construction of the final tree that we will return */
+  final_tree = kvtree_new();
+
+  /* Look at all the subfiles with our prefix and read in the file lists */
+  prefix_dir_copy = strdup(prefix);
+  prefix_dir = dirname(prefix_dir_copy);
+  prefix_file_copy = strdup(prefix);
+  prefix_file = basename(prefix_file_copy);
+
+  DIR *d;
+  struct dirent *dir;
+
+  d = opendir(prefix_dir);
+  if (!d) {
+    rc = KVTREE_FAILURE;
+    goto end;
+  }
+
+  snprintf(pattern, sizeof(pattern), "^%s\\.0\\.[0-9]+", prefix_file);
+  pattern[sizeof(pattern) - 1] = '\0';
+  rc = regcomp(&regex, pattern, REG_EXTENDED);
+  if (rc) {
+    printf("%s: couldn't compile regex\n", __func__);
+    rc = KVTREE_FAILURE;
+    goto end;
```

Comment:
As this is ordered, we need to close the directory before returning due to error.  Instead, let's move this block to come before the opendir, as that's one less line and it moves the opendir to be closer to the readdir/closedir calls.